### ISSUE TITLE
String replace regex makes all the following lines as a string

### ISSUE THREAD

**Issue by hexa00** [ananthakumaran/typescript.el#34]
Something like:

test.replace(/\\/g,"/");
nextlines are as if part of a string

Using : typescript-mode-20170324.1301

**Comment by lddubeau** [comment]
The mode is working correctly. It is your regex that is wrong and throwing you off. 

You expect the first `/` to terminate your regular expression, but you have `\` before it, which escapes its normal meaning. You think your regular expression is:

```
/\/g
```

but it is:

```
/\/g,"/
```

and then it is followed by the start of a string which you never terminate.

The regex you want appears to be this: `/\\/g`. This will escape the escape character and make it match a backslash.

**Comment by hexa00** [comment]
oops seems like the github comment missed a \ here... I should have enclosed it

The real code is like you said:
`test.replace(/\\/g,"/");`
With the same result...

**Comment by lddubeau** [comment]
Ah, yes, I can reproduce it here now.

**Comment by hexa00** [comment]
:) Note I tested in the JavaScript major mode and it works there...

**Comment by josteink** [comment]
I did some basic testing and landed on this:

````
test.replace(/a/g,"/");    // good
test.replace(/a\a/g,"/");  // good
test.replace(/a\\a/g,"/"); // good
test.replace(/a\\/g,"/");  // not good
````

Basically if we have a `\` of any sort preceede the finalizing `/`, our code breaks. Not having looked at the code in question, it seems like some naive escape-handling code fails to realize detect the double `\\` as one escaped character, and thinks we're trying to escape the `/`.

**Comment by sudarshang** [comment]
Looking into how js-mode handles this

https://github.com/emacs-mirror/emacs/blob/master/lisp/progmodes/js.el#L1725
```elisp
(defun js-syntax-propertize-regexp (end)
  (let ((ppss (syntax-ppss)))
    (when (eq (nth 3 ppss) ?/)
      ;; A /.../ regexp.
      (goto-char (nth 8 ppss))
      (when (looking-at js--syntax-propertize-regexp-regexp)
        ;; Don't touch text after END.
        (when (> end (match-end 1))
          (setq end (match-end 1)))
        (put-text-property (match-beginning 1) end
                           'syntax-table (string-to-syntax "\"/"))
        (goto-char end)))))
```

It seems to use a function to deal with regexes instead of using the regular expression that typescript mode uses

https://github.com/ananthakumaran/typescript.el/blob/master/typescript-mode.el#L1624
```elisp
(defconst typescript--regexp-literal
  "[=(,:]\\(?:\\s-\\|\n\\)*\\(/\\)\\(?:\\\\/\\|[^/*]\\)\\(?:\\\\/\\|[^/]\\)*\\(/\\)"
  "Regexp matching a typescript regular expression literal.
Match groups 1 and 2 are the characters forming the beginning and
end of the literal.")
```
Is switching to how js mode handles regexes an option for typescript mode?

**Comment by sudarshang** [comment]
Looks like typescript mode was forked from a much older version of js-mode. Are there benefits to grabbing the current js-mode and trying to incorporate all the fixes that were made to it since then? I would be interested in taking this on if the current maintainers 
1. Think this would be useful.
2. Would be willing to do a lot of handholding and question answering

**Comment by josteink** [comment]
On one side I think that would be interesting. Having a more unified code base *could* be nice and might let us solve bugs like this more easily. 

On the other hand, @ananthakumaran picked up this code from someone else (and admits he doesn't fully get it) and none of the original developers are around for the "hand holding" bit. I honestly think there's a limited amount of help we can provide. 

Take a look and see how big the job looks?

**Comment by sudarshang** [comment]
I checked out old versions of js-mode.el (git tag emacs-pretest-24.0.95) and typescript-mode.el. (git hash 09123ba). I then did a global replace of typescript with js to fix the spurious diffs introduced by renaming js to typescript. Figuring out what was added to js-mode to get the initial typescript mode leads to the following file

https://gist.github.com/sudarshang/f37d22d4c580080f5c103e2f6e4c8a31

I am going to try making typescript mode a derived mode of js-mode next, do let me know if that sounds like a bad idea.

**Comment by nspaeth** [comment]
@sudarshang That seems like it might be a lot of work. 
Would it be easier or better to start making a typescript  [lsp-mode](https://github.com/emacs-lsp/lsp-mode)?

**Comment by josteink** [comment]
The parts covered by lsp-mode is already handled by the tide-package. 

When lsp-mode works satisfactory well, tide may be obsolete, but grammar, font-locking, indentation and stuff like that will still need to be handled by Emacs. 

I.e. the stuff covered by typescript-mode. This should be fixed here.

**Comment by sudarshang** [comment]
PR [45](https://github.com/ananthakumaran/typescript.el/pull/45) does fix this issue.

**Comment by josteink** [comment]
With PR #45 closed (and not merged), we'll have to come up with a solution based on our existing code.

**Comment by bkruczyk** [comment]
Does anyone see a reason why string literal for regexp can't be simplified to `"[=(,:]\\(?:\\s-\\|\n\\)*\\(/\\)[^/]*\\(/\\)"`. Check PR #55.

**Comment by josteink** [comment]
Just for clarity for other looking into this thread: Since @bkruczyk has already a PR submitted, I've decided to take technical comments regarding his suggestion in the PR, and not this issue.

**Comment by josteink** [comment]
Fixed in #55.
